libraries/aws/shadow/src/shadow.c

```diff
@@ -0,0 +1,550 @@
+                                                 pMessageType );
+        if( shadowStatus != SHADOW_STATUS_SUCCESS )
+        {
+            LogDebug( ( "Not related to Shadow, failed to match shadow message type in pTopic %s", pTopic ) );
+        }
+    }
+
+    return shadowStatus;
+}
+/*-----------------------------------------------------------*/
+
+ShadowStatus_t Shadow_GetTopicString( ShadowTopicStringType_t topicType,
+                                      const char * pThingName,
+                                      uint8_t thingNameLength,
+                                      char * pTopicBuffer,
+                                      uint16_t bufferSize,
+                                      uint16_t * pOutLength )
+{
+    uint16_t offset = 0U, generatedTopicStringLength = 0U;
+    ShadowStatus_t shadowStatus = SHADOW_STATUS_SUCCESS;
+
+    /* Lookup table for Shadow operation string. */
+    static const char * const pTopicString[ SHADOW_TOPIC_STRING_TYPE_MAX_NUM ] =
+    {
+        SHADOW_TOPIC_OPERATION_STRING_GET,
+        SHADOW_TOPIC_OPERATION_STRING_GET_ACCEPTED,
+        SHADOW_TOPIC_OPERATION_STRING_GET_REJECTED,
+        SHADOW_TOPIC_OPERATION_STRING_DELETE,
+        SHADOW_TOPIC_OPERATION_STRING_DELETE_ACCEPTED,
+        SHADOW_TOPIC_OPERATION_STRING_DELETE_REJECTED,
+        SHADOW_TOPIC_OPERATION_STRING_UPDATE,
+        SHADOW_TOPIC_OPERATION_STRING_UPDATE_ACCEPTED,
+        SHADOW_TOPIC_OPERATION_STRING_UPDATE_REJECTED,
+        SHADOW_TOPIC_OPERATION_STRING_UPDATE_DOCUMENTS,
+        SHADOW_TOPIC_OPERATION_STRING_UPDATE_DELTA
+    };
+
+    /* Lookup table for Shadow operation string length. */
+    static const uint16_t pTopicStringLength[ SHADOW_TOPIC_STRING_TYPE_MAX_NUM ] =
+    {
+        SHADOW_TOPIC_OPERATION_LENGTH_GET,
+        SHADOW_TOPIC_OPERATION_LENGTH_GET_ACCEPTED,
+        SHADOW_TOPIC_OPERATION_LENGTH_GET_REJECTED,
+        SHADOW_TOPIC_OPERATION_LENGTH_DELETE,
+        SHADOW_TOPIC_OPERATION_LENGTH_DELETE_ACCEPTED,
+        SHADOW_TOPIC_OPERATION_LENGTH_DELETE_REJECTED,
+        SHADOW_TOPIC_OPERATION_LENGTH_UPDATE,
+        SHADOW_TOPIC_OPERATION_LENGTH_UPDATE_ACCEPTED,
+        SHADOW_TOPIC_OPERATION_LENGTH_UPDATE_REJECTED,
+        SHADOW_TOPIC_OPERATION_LENGTH_UPDATE_DOCUMENTS,
+        SHADOW_TOPIC_OPERATION_LENGTH_UPDATE_DELTA
+    };
+
+    if( ( pTopicBuffer == NULL ) ||
+        ( pThingName == NULL ) ||
+        ( thingNameLength == 0U ) ||
+        ( topicType >= SHADOW_TOPIC_STRING_TYPE_MAX_NUM ) ||
+        ( pOutLength == NULL ) )
+    {
+        shadowStatus = SHADOW_STATUS_BAD_PARAMETER;
+        LogError( ( "Invalid input parameters" ) );
+    }
+    else
+    {
+        generatedTopicStringLength = SHADOW_TOPIC_PREFIX_LENGTH +       /* Prefix ("$aws/things/"). */
+                                     thingNameLength +                  /* Thing name. */
+                                     pTopicStringLength[ topicType ];   /* Shadow operation. */
```

Comment:
I'm not sure I like this method of indexing into the array with the value of the enum. If we ever reorder the enums in the header file, then we'd need to remember to reorder them wherever these arrays are used as well. Why not create a `getStringFromEnum` function that returns the values used in the array with a `switch` statement?

Reply:
This is following the pattern from Shadow v4_beta
https://github.com/aws/aws-iot-device-sdk-embedded-C/blob/v4_beta/libraries/aws/shadow/src/aws_iot_shadow_api.c#L474

And also I'm sure if it is worth it to write functions to get the length, strings for each case.
Let me know if you still suggest to change it

Thank you

Reply:
I was thinking you'd have a function like:
```
void getTopicStringAndLengthFromType( ShadowTopicStringType_t topicType,
                                 uint16_t * pLength,
                                 const char ** pString )
{
    switch( topicType )
    {
        case ShadowTopicStringTypeGet:
            *pString = SHADOW_TOPIC_OPERATION_STRING_GET;
            *pLength = SHADOW_TOPIC_OPERATION_LENGTH_GET;
            break;
        case ShadowTopicStringTypeGetAccepted:
        ...
        /* Other cases. */
        ...
        default:
            break;
    }
}
```
Then, to reuse this function in `extractShadowMessageType`, you'd only need to use a single array:
```
    static const ShadowTopicStringType_t pMessageTypes[ ShadowMessageTypeMaxNum ] =
    {
        ShadowTopicStringTypeGetAccepted,
        ShadowTopicStringTypeGetRejected,
        ShadowTopicStringTypeDeleteAccepted,
        ShadowTopicStringTypeDeleteRejected,
        ShadowTopicStringTypeUpdateAccepted,
        ShadowTopicStringTypeUpdateRejected,
        ShadowTopicStringTypeUpdateDocument,
        ShadowTopicStringTypeUpdateDelta
    };
```
Both functions would call `getTopicStringAndLengthFromType`; `extractShadowMessageType` would only need one array where ordering doesn't matter, and this function wouldn't need any arrays. Do you think this would be useful?

Reply:
I agree with Muneeb and we should add a helper function to avoid depending on the enum values for indexing.

Reply:
fixed it in commit [8da68c5](https://github.com/aws/aws-iot-device-sdk-embedded-C/pull/1081/commits/8da68c5f3ccc011a9ab8f58349cf6273c42ec19c)

by added getShadowOperationString, getShadowOperationLength and removed indexing into the array with the value of the enum as suggested.

resolving this comment

feel free to let me know if it needs further changes, thank you